server/conn.go

```diff
@@ -2156,11 +2156,43 @@ func (cc getLastStmtInConn) PProfLabel() string {
 	case mysql.ComStmtReset:
 		return "ResetStmt"
 	case mysql.ComQuery, mysql.ComStmtPrepare:
-		return parser.Normalize(queryStrForLog(string(hack.String(data))))
+		return parser.Normalize(tidbutil.QueryStrForLog(string(hack.String(data))))
 	case mysql.ComStmtExecute, mysql.ComStmtFetch:
 		stmtID := binary.LittleEndian.Uint32(data[0:4])
-		return queryStrForLog(cc.preparedStmt2StringNoArgs(stmtID))
+		return tidbutil.QueryStrForLog(cc.preparedStmt2StringNoArgs(stmtID))
 	default:
 		return ""
 	}
 }
+
+// PProfLabelNormalizedAndDigest return sql and sql_digest label used to tag pprof.
+func (cc getLastStmtInConn) pprofLabelNormalizedAndDigest() (string, string) {
+	if len(cc.lastPacket) == 0 {
+		return "", ""
+	}
+	cmd, data := cc.lastPacket[0], cc.lastPacket[1:]
+	switch cmd {
+	case mysql.ComInitDB:
+		return "UseDB", ""
+	case mysql.ComFieldList:
+		return "ListFields", ""
+	case mysql.ComStmtClose:
+		return "CloseStmt", ""
+	case mysql.ComStmtReset:
+		return "ResetStmt", ""
+	case mysql.ComQuery, mysql.ComStmtPrepare:
+		normalized, digest := parser.NormalizeDigest(string(hack.String(data)))
+		return normalized, digest.String()
+	case mysql.ComStmtExecute, mysql.ComStmtFetch:
+		stmtID := binary.LittleEndian.Uint32(data[0:4])
+		prepareObj, _ := cc.preparedStmtID2CachePreparedStmt(stmtID)
+		if prepareObj != nil && prepareObj.SQLDigest != nil {
+			return prepareObj.NormalizedSQL, prepareObj.SQLDigest.String()
+		}
+		str := cc.preparedStmt2StringNoArgs(stmtID)
+		normalized, digest := parser.NormalizeDigest(str)
```

Comment:
This introduces each preparedStatement execution two extra cost:

1. serialize prepared statement back to string
2. parse string to calculate a digest

Reply:
No, for the normal preparedStatement, it should return earlier. 
Run to here indicate the prepared statements can't be found or invalid, so it must be somewhere wrong.

Reply:
Then maybe we should print or returning errors? In what kind of scenario `prepareObj == nil` and execution can be success?

BTW seems that prepared statements are stored in a "cache", which means items inside can be just dropped naturally?

Reply:
Already remove this, PTAL